pdf/creator/styled_paragraph.go

```diff
@@ -0,0 +1,629 @@
+/*
+ * This file is subject to the terms and conditions defined in
+ * file 'LICENSE.md', which is part of this source code package.
+ */
+
+package creator
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"unicode"
+
+	"github.com/unidoc/unidoc/common"
+	"github.com/unidoc/unidoc/pdf/contentstream"
+	"github.com/unidoc/unidoc/pdf/core"
+	"github.com/unidoc/unidoc/pdf/model/textencoding"
+)
+
+// StyledParagraph represents text drawn with a specified font and can wrap across lines and pages.
+// By default occupies the available width in the drawing context.
+type StyledParagraph struct {
+	// Text chunks with styles that compose the paragraph
+	chunks []TextChunk
+
+	// Style used for the paragraph for spacing and offsets
+	defaultStyle TextStyle
+
+	// The text encoder which can convert the text (as runes) into a series of glyphs and get character metrics.
+	encoder textencoding.TextEncoder
+
+	// Text alignment: Align left/right/center/justify.
+	alignment TextAlignment
+
+	// The line relative height (default 1).
+	lineHeight float64
+
+	// Wrapping properties.
+	enableWrap bool
+	wrapWidth  float64
+
+	// defaultWrap defines whether wrapping has been defined explictly or whether default behavior should
+	// be observed. Default behavior depends on context: normally wrap is expected, except for example in
+	// table cells wrapping is off by default.
+	defaultWrap bool
+
+	// Rotation angle (degrees).
+	angle float64
+
+	// Margins to be applied around the block when drawing on Page.
+	margins margins
+
+	// Positioning: relative / absolute.
+	positioning positioning
+
+	// Absolute coordinates (when in absolute mode).
+	xPos float64
+	yPos float64
+
+	// Scaling factors (1 default).
+	scaleX float64
+	scaleY float64
+
+	// Text chunk lines after wrapping to available width.
+	lines [][]TextChunk
+}
+
+// NewStyledParagraph create a new styled paragraph.
+// Uses default parameters: Helvetica, WinAnsiEncoding and wrap enabled
+// with a wrap width of 100 points.
+func NewStyledParagraph(text string, style TextStyle) *StyledParagraph {
+	// TODO: Can we wrap intellectually, only if given width is known?
+	p := &StyledParagraph{
+		chunks: []TextChunk{
+			TextChunk{
+				Text:  text,
+				Style: style,
+			},
+		},
+		defaultStyle: NewTextStyle(),
```

Comment:
The style argument could theoretically be used there instead of creating a new style. However, the default style is used to create the offsets (shifts) for the text alignments (center/right) and also for rendering spaces when the text alignment is justify. I think it's safer we use a predictable style instead of accepting any font the user can provide in the style argument for this.